Fix #210, Add SB API to get pipe by name

CDKnightNASA opened the pull request:
**Describe the contribution**
Fixes issue #210 

**Testing performed**
make mission-all; run cpu1/cfe_core_default_cpu1/unit-test/cfe_core_default_cpu1_sb_UT

**Expected behavior changes**
Adds a new function, CFE_SB_GetPipeIdByName, which retrieves the pipe ID given a name of a pipe. As this would conflict with the removal of the PipeName from the PipeTbl, this code uses the OS_QueueGetIdByName and iterates across the PipeTbl to find the matching entry.

**System(s) tested on:**
Ubuntu 64-bit Linux 19.04.

**Contributor Info**
Chris Knight, NASA Ames Research Center.

skliper commented:
CCB 20191113 - pipe id and queue id are of different sizes/types, could look into collapsing into 1 type, associated with the type safe ticket also, could also optimize by mapping them 1 to 1.

skliper commented:
CCB 20191123 - approved for integration candidate, under additional review by jphickey and acudmore

acudmore commented on fsw/cfe-core/src/inc/cfe_sb.h:
```diff
@@ -267,11 +267,32 @@ int32  CFE_SB_SetPipeOpts(CFE_SB_PipeId_t     PipeId,
+** \returns
+** \retcode #CFE_SUCCESS           \retdesc  \copydoc CFE_SUCCESS         \endcode
+** \retcode #CFE_SB_BAD_ARGUMENT   \retdesc  \copydoc CFE_SB_BAD_ARGUMENT \endcode
+** \endreturns
```
Also add CFE_SB_INVALID_PIPE here?

acudmore commented on fsw/cfe-core/src/inc/cfe_sb_msg.h:
```diff
@@ -559,6 +559,8 @@ typedef struct {
+                                        \brief Count of errors in get pipe id by name API */
     uint8           DuplicateSubscriptionsCounter;/**< \cfetlmmnemonic \SB_DUPSUBCNT
                                              \brief Count of duplicate subscriptions */
     uint8           Spare2Align[2];/**< \cfetlmmnemonic \SB_SPARE2ALIGN
```
Does adding a byte of telemetry change the Spare2Align bytes?

CDKnightNASA commented:
> Also, since `CFE_SB_GetPipeIdByName` is being added to the public SB API, it needs a stub function implemented in `ut_sb_stubs.c` too. I'd suggest using `CFE_ES_GetAppIdByName` stub as a guide but that one is more complex than necessary (I think it had to support some existing oddball test scenarios). It should be just as simple as using the UT_DEFAULT_IMPL macro followed up by a UT_Stub_CopyToLocal() to allow a test case to provide its own outputs.

Looking at the CFE_ES_GetAppIDByName, I realize there is no CFE_SB_GetPipeName(char *buf, PipeId_t id) ...should I add it?

Also while a bit nit-picky, it's good to resolve these things for a style guide. I note two things:

1) ES uses "AppID" most places, not "AppId" in names of functions. SB uses PipeId. We should decide which is "suggested style" for our style guide.

2) I'm not fond of GetFoo(char *OutBuf, int inId, size_t BufLen), would prefer the len variable be adjacent to the outbuf variable, so GetFoo(char *OutBuf, size_t BufLen, int inId.)

I've captured both in my draft "style" doc (in the fix-404-contributing branch) as "to be decided." Along with other TBD's.

skliper commented:
CCB - 12/4/2019: Consider adding an OSAL API which just retrieves the name of a resource, simplifies this issue and the other's related

CDKnightNASA commented:
I need to squash

skliper commented:
Is this ready for merge to integration branch?  I don't see all the requests being closed out.  @acudmore  @jphickey

jphickey commented:
I would still like to see this squashed and rebased into a single commit.  There is lots of back and forth on the branch now (incl. some bogus/mistake commits).  This should be cleaned up before merging.

CDKnightNASA commented:
I’ve reviewed the requests and confirmed all have been addressed (except for resolving the squash, see my other e-mail a minute ago.)

From: skliper <notifications@github.com>
Sent: Tuesday, January 7, 2020 11:24
To: nasa/cFE <cFE@noreply.github.com>
Cc: Knight, Christopher D. (ARC-TI) <christopher.d.knight@nasa.gov>; Author <author@noreply.github.com>
Subject: [EXTERNAL] Re: [nasa/cFE] Fix #210, Add SB API to get pipe by name (#404)


Is this ready for merge to integration branch? I don't see all the requests being closed out. @acudmore<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_acudmore&d=DwMCaQ&c=ApwzowJNAKKw3xye91w7BE1XMRKi2LN9kiMk5Csz9Zk&r=JDJB_KO_kIJMmlWcOl5Br6wthfeLCcxzUseDoBjuvfFQOIpFSnwdteHjX4mn9USx&m=QXpwTyPhiJJrS64tP_eWarxv6CLW2pdqh1iU-QBGHn0&s=dVgv-2dteDGEHbO_1d5uUgXAvslxwDwoHCxfmM6Qu8M&e=> @jphickey<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_jphickey&d=DwMCaQ&c=ApwzowJNAKKw3xye91w7BE1XMRKi2LN9kiMk5Csz9Zk&r=JDJB_KO_kIJMmlWcOl5Br6wthfeLCcxzUseDoBjuvfFQOIpFSnwdteHjX4mn9USx&m=QXpwTyPhiJJrS64tP_eWarxv6CLW2pdqh1iU-QBGHn0&s=AmM3pPbfPzlkF2PuVEnCg3D94_FWKOfCMn_h0CQ2zxE&e=>

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_nasa_cFE_pull_404-3Femail-5Fsource-3Dnotifications-26email-5Ftoken-3DAIZYMK6WQORSRZVUY3L34K3Q4TJE7A5CNFSM4JMOLHK2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIJ7HMQ-23issuecomment-2D571732914&d=DwMCaQ&c=ApwzowJNAKKw3xye91w7BE1XMRKi2LN9kiMk5Csz9Zk&r=JDJB_KO_kIJMmlWcOl5Br6wthfeLCcxzUseDoBjuvfFQOIpFSnwdteHjX4mn9USx&m=QXpwTyPhiJJrS64tP_eWarxv6CLW2pdqh1iU-QBGHn0&s=bl-MPgyI24ZQE5zZ45Yq2VBExwzV9fyl6M8GqvFk2m4&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AIZYMK3GO6ICF63VFBS7CZ3Q4TJE7ANCNFSM4JMOLHKQ&d=DwMCaQ&c=ApwzowJNAKKw3xye91w7BE1XMRKi2LN9kiMk5Csz9Zk&r=JDJB_KO_kIJMmlWcOl5Br6wthfeLCcxzUseDoBjuvfFQOIpFSnwdteHjX4mn9USx&m=QXpwTyPhiJJrS64tP_eWarxv6CLW2pdqh1iU-QBGHn0&s=WoRd-hTGdiI-i8FBmwPTUdxsqh-MpXX38H0xtJGUZ3M&e=>.

skliper commented:
I rebased on current master and squashed for review.  @acudmore and @jphickey please re-review.

jphickey commented on fsw/cfe-core/src/sb/cfe_sb_api.c:
```diff
@@ -540,6 +536,169 @@ int32 CFE_SB_GetPipeOpts(CFE_SB_PipeId_t PipeId, uint8 *OptsPtr)
+    }else{
+        if (OS_QueueGetInfo(CFE_SB.PipeTbl[PipeId].SysQueueId, &queue_prop)
+                == OS_SUCCESS){
+            strncpy(PipeNameBuf, queue_prop.name, OS_MAX_API_NAME-1);
```
This needs to use `PipeNameSize`, not OS_MAX_API_NAME-1.  It should also ensure a null-terminated output.

CDKnightNASA replied:
fixed

jphickey replied:
Hate to be a pain, but still need to make sure that the output is null terminated.  This should also handle the case if the passed in size is 0 (since it is handling a passed-in buffer of NULL, should handle this too).

jphickey commented on fsw/cfe-core/src/sb/cfe_sb_api.c:
```diff
@@ -704,6 +863,9 @@ int32  CFE_SB_SubscribeFull(CFE_SB_MsgId_t   MsgId,
     char   FullName[(OS_MAX_API_NAME * 2)];
+    char   PipeName[OS_MAX_API_NAME] = {'\0'};
+
+    CFE_SB_GetPipeName(PipeName, OS_MAX_API_NAME, PipeId);
```
I prefer to see `sizeof(PipeName)` rather than `OS_MAX_API_NAME`, as this allows the string definition to change without breaking the code.

CDKnightNASA replied:
resolved

jphickey commented on fsw/cfe-core/src/sb/cfe_sb_task.c:
```diff
@@ -898,7 +898,7 @@ int32 CFE_SB_SendRtgInfo(const char *Filename)
                  */
                 CFE_ES_GetAppName(&Entry.AppName[0], pd->AppId, sizeof(Entry.AppName));
-                strncpy(&Entry.PipeName[0],CFE_SB_GetPipeName(Entry.PipeId),sizeof(Entry.PipeName));
+                CFE_SB_GetPipeName(Entry.PipeName, OS_MAX_API_NAME, Entry.PipeId);
```
Should continue to use `sizeof(Entry.PipeName)` as the original did.

skliper commented:
Analyzing log I see I should have done the squash differently (new method used successfully on https://github.com/nasa/cFE/pull/452) to retain authorship by Chris.

CDKnightNASA commented:
I’ve been spending the morning getting comfortable with git reset in prep to do this squash, but you beat me to it. :)

From: skliper <notifications@github.com>
Sent: Tuesday, January 14, 2020 12:38
To: nasa/cFE <cFE@noreply.github.com>
Cc: Knight, Christopher D. (ARC-TI) <christopher.d.knight@nasa.gov>; Author <author@noreply.github.com>
Subject: [EXTERNAL] Re: [nasa/cFE] Fix #210, Add SB API to get pipe by name (#404)


I rebased on current master and squashed for review. @acudmore<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_acudmore&d=DwMCaQ&c=ApwzowJNAKKw3xye91w7BE1XMRKi2LN9kiMk5Csz9Zk&r=JDJB_KO_kIJMmlWcOl5Br6wthfeLCcxzUseDoBjuvfFQOIpFSnwdteHjX4mn9USx&m=pZr_5tgbpe3sbgz4i8JiCGWcZxhhV30U0ebGXAby7Fo&s=dxWds1usA1P1FMPM1U0NvdYjqjqG2HXptKC0Kd06b0s&e=> and @jphickey<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_jphickey&d=DwMCaQ&c=ApwzowJNAKKw3xye91w7BE1XMRKi2LN9kiMk5Csz9Zk&r=JDJB_KO_kIJMmlWcOl5Br6wthfeLCcxzUseDoBjuvfFQOIpFSnwdteHjX4mn9USx&m=pZr_5tgbpe3sbgz4i8JiCGWcZxhhV30U0ebGXAby7Fo&s=rbXNb87qj8n-bfSDgWH4sBUd2RUACkiYAQ4rBqphaa0&e=> please re-review.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_nasa_cFE_pull_404-3Femail-5Fsource-3Dnotifications-26email-5Ftoken-3DAIZYMKYJJJD6H25MVEDA55LQ5YPDFA5CNFSM4JMOLHK2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEI6BIGY-23issuecomment-2D574362651&d=DwMCaQ&c=ApwzowJNAKKw3xye91w7BE1XMRKi2LN9kiMk5Csz9Zk&r=JDJB_KO_kIJMmlWcOl5Br6wthfeLCcxzUseDoBjuvfFQOIpFSnwdteHjX4mn9USx&m=pZr_5tgbpe3sbgz4i8JiCGWcZxhhV30U0ebGXAby7Fo&s=mE-_8l0vSog1CjCimBKlUfYJVXeBSNFQxqnHu4-l9KI&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AIZYMK37AW4DCRAQUTCSWRLQ5YPDFANCNFSM4JMOLHKQ&d=DwMCaQ&c=ApwzowJNAKKw3xye91w7BE1XMRKi2LN9kiMk5Csz9Zk&r=JDJB_KO_kIJMmlWcOl5Br6wthfeLCcxzUseDoBjuvfFQOIpFSnwdteHjX4mn9USx&m=pZr_5tgbpe3sbgz4i8JiCGWcZxhhV30U0ebGXAby7Fo&s=W-QaGXCdRxGONxxIEqlAR3L0YGleLdPndctpt5S_YcQ&e=>.

CDKnightNASA commented:
re-squashed

skliper commented:
CCB 20200122 - Reviewed and approved for CI
